flytekit/core/type_engine.py

```diff
@@ -481,6 +482,7 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
         dc = cast(DataClassJsonMixin, expected_python_type).from_json(_json_format.MessageToJson(lv.scalar.generic))
         return self._fix_dataclass_int(expected_python_type, self._deserialize_flyte_type(dc, expected_python_type))
 
+    @lru_cache(typed=True)
```

Comment:
added this since the `run` command needs to call `guess_python_type` to get the TypeEngine-derived dataclass. Without caching here, a separate call to `guess_python_type` in the `remote.execute` call would result in a logically equivalent (but new) dataclass, which `TypeEngine.assert_type` would not be happy about.